File: consistency-store/src/main/java/de/tudarmstadt/consistency/store/local/MapStore.java
package de.tudarmstadt.consistency.store.local;

import de.tudarmstadt.consistency.store.Store;
import de.tudarmstadt.consistency.store.Transaction;
import org.checkerframework.com.google.common.collect.Maps;

import java.util.Map;

/**
 * Created on 24.05.18.
 *
 * @author Mirko Köhler
 */
public class MapStore implements Store<Object, MapReferenceService> {

	private final Map<Object, Object> data;

	private MapStore() {
		data = Maps.newHashMap();
	}

	public static MapStore create() {
		return new MapStore();
	}

	@Override
	public void commit(Transaction<MapReferenceService> actions, Class<?> isolationLevel) throws Exception {
		actions.executeWith(new MapReferenceService(this));
	}

	void put(Object id, Object value) {
		data.put(id, value);
	}

	Object get(Object id) {
		return data.getOrDefault(id, null);
	}


}

File: consistency-store/src/main/scala/de/tudarmstadt/consistency/store/scala/local/MapStore.scala
package de.tudarmstadt.consistency.store.scala.local

import java.lang.annotation.Annotation

import de.tudarmstadt.consistency.checker.qual.{Strong, Weak}
import de.tudarmstadt.consistency.store.scala.impl.{ReadWriteRef, ReadWriteStore}

import scala.collection.mutable

/**
	* Created on 09.08.18.
	*
	* @author Mirko Köhler
	*/
class MapStore[Key, Val] extends ReadWriteStore[Key, Val] {

	override type Context = MapSessionContext

	override def newSessionContext(): Context = new MapSessionContext


	private val data : mutable.HashMap[Key, Any] = mutable.HashMap.empty

	/**
		* Session contexts when working with maps.
		*/
	class MapSessionContext extends ReadWriteSessionContext {

		override def obtain[T <: Val](key: Key, consistencyLevel : Class[_ <: Annotation]): ReadWriteRef[Key, T] = consistencyLevel match {
			case x if classOf[Weak] == x => new MapRef(data, key)
			case x if classOf[Strong] == x => new MapRef(data, key)
			case x => throw new IllegalArgumentException(s"unsupported consistency level. expected Weak or Strong, but got $x")
		}
	}


}



